lib/segment/src/types.rs

```diff
@@ -874,14 +874,47 @@ pub struct Payload(pub Map<String, Value>);
 
 impl Payload {
     pub fn merge(&mut self, value: &Payload) {
-        for (key, value) in &value.0 {
+        Self::merge_map(&mut self.0, value)
+    }
+
+    pub fn merge_by_key(&mut self, value: &Payload, key: &str) -> OperationResult<()> {
+        let v = self.get_object_mut(key)?;
+        Self::merge_map(v, value);
+        Ok(())
+    }
+
+    fn merge_map(dest: &mut Map<String, Value>, source: &Payload) {
+        for (key, value) in &source.0 {
             match value {
-                Value::Null => self.0.remove(key),
-                _ => self.0.insert(key.to_owned(), value.to_owned()),
+                Value::Null => dest.remove(key),
+                _ => dest.insert(key.to_owned(), value.to_owned()),
             };
         }
     }
 
+    fn get_object_mut(&mut self, key: &str) -> OperationResult<&mut Map<String, Value>> {
+        let segs = key.split('.').collect::<Vec<&str>>();
+        if segs.is_empty() {
+            return Err(OperationError::WrongPayloadKey {
+                description: format!("Unexpected payload key: \"{key}\""),
+            });
+        }
+        let mut v = &mut self.0;
+        for seg in segs {
+            let v1 = v.get_mut(seg).ok_or(OperationError::WrongPayloadKey {
+                description: format!("Can't find payload key segment: \"{seg}\"."),
```

Comment:
I think this should create required nested object instead of faiting with error.